book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+    - Remove anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+    - Subtract work from `self.partial_cumulative_work`
+
+3. Return the block
+
+#### `Ord`
+
+The `Chain` type also implements `Ord` for reorganizing chains. First chains
+are compared by their `partial_cummulative_work`. Ties are then broken by
+comparing `block::Hash`es of the tips of each chain.
+
+**Note**: Unlike `zcashd`, Zebra does not use block arrival times as a tie-breaker for the best tip. Since Zebra downloads blocks in parallel, download times are not guaranteed to be unique. Using the `block::Hash` provides a consistent tip order. (As a side-effect, the tip order is also consistent after a node restart, and between nodes.)
+### `ChainSet` Type
+[chainset-type]: #chainset-type
+
+The `ChainSet` type represents the set of all non-finalized state. It
+consists of a set of non-finalized but verified chains and a set of
+unverified blocks which are waiting for the full context needed to verify
+them to become available.
+
+`ChainSet` is defined by the following structure and API:
+
+```rust
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
+```
+
+#### `pub fn finalize(&mut self) -> Arc<Block>`
+
+Finalize the lowest height block in the non-finalized portion of the best
+chain and updates all side chains to match.
+
+1. Extract the best chain from `self.chains` into `best_chain`
+
+2. Extract the rest of the chains into a `side_chains` temporary variable, so
+   they can be mutated
+
+3. Remove the lowest height block from the best chain with
+   `let block = best_chain.pop_root();`
+
+4. Add `best_chain` back to `self.chains`
+
+5. For each remaining `chain` in `side_chains`
+    - If `chain` starts with `block`, remove `block` and add `chain` back to
+    `self.chains`
+    - Else, drop `chain`
+
+6. for each `height` in `self.queued_by_height` where the height is lower than the
+   new reorg limit
+   - for each `hash` in `self.queued_by_height.remove(height)`
+     - Remove the key `hash` from `self.queued_blocks` and store the removed `block`
+     - Find and remove `hash` from `self.queued_by_parent` using `block.parent`'s hash
+
+7. Return `block`
+
+### `pub fn queue(&mut self, block: QueuedBlock)`
+
+Queue a non-finalized block to be committed to the state.
+
+After queueing a non-finalized block, this method checks whether the newly
+queued block (and any of its descendants) can be committed to the state
+
+1. Check if the parent block exists in any current chain
+
+2. If it does, call `let ret = self.commit_block(block)`
+    - Call `self.process_queued(new_parents)` if `ret` is `Some`
+
+3. Else Add `block` to `self.queued_blocks` and related members and return
```

Comment:
```suggestion
1. If the parent block exists in any current chain
  - call `let ret = self.commit_block(block)`
  - if `ret` is `Some`
    - call `self.process_queued(new_parents)` 
  - otherwise, drop `block`, because it can never be verified

2. Else Add `block` to `self.queued_blocks` and related members and return
```

Reply:
> If a block's parent is present, but it can't be verified, we'll never be able to verify it.

we specified `commit_block` to panic if called on a block that can't be verified to handle this case

Reply:
How do we handle duplicates, @yaahc?

Reply:
> > If a block's parent is present, but it can't be verified, we'll never be able to verify it.
> 
> we specified `commit_block` to panic if called on a block that can't be verified to handle this case

This answer doesn't make sense to me, because `commit_block` won't be called until the parent is received.

Reply:
Ah, I see, these parts of the spec were very hard for me to read, because they still contained the old `Option<block::Hash>` return value, and some outdated error handling.

See https://github.com/ZcashFoundation/zebra/pull/1090/files#diff-0a5b73420cea8c2ce73f46bc2b5d56c7R446 for some changes that should make this a lot clearer.

Reply:
> How do we handle duplicates, @yaahc?

See https://github.com/ZcashFoundation/zebra/pull/1090/files#diff-0a5b73420cea8c2ce73f46bc2b5d56c7R425 , where I suggested the duplicate-handling change again.